### PR TITLE
Fixes the subs(x, 0).subs(x, y) problem

### PR DESCRIPTION
**Pull request by certik**

### CODE REVIEW

**Comment by certik** [comment]
@smichr you wrote the original code of `Subs._eval_subs()` in https://github.com/sympy/sympy/commit/aac30a2ba728ff6980266ed6a469278eaf8ae575#diff-df6e27def8b4f219ead0c9b48a032d66L1446. Would you mind reviewing this PR?

**Comment by asmeurer** [comment]
The second argument to subs is a bound variable, so subs shouldn't affect it.

**Comment by asmeurer** [comment]
The way I see it, the `x` in `Subs(f(x), x, 0)` is a bound symbol (it's the same as `Subs(f(y), y, 0)`). There seems to be two classes of behavior with this, one is to ignore it with subs (this is what Integral does, like `Integral(f(x), (x, 0, 1)).subs(x, y)`, and another is to just swap out the symbol (like with `Lambda(x, f(x))`). 

I seem to remember a lot of discussion on the pros and cons of each approach in some issue. Either one is probably fine, but the current behavior of Subs seems wrong.

File: sympy/core/function.py [hunk 1]
```diff
@@ -1556,9 +1556,7 @@ def _hashable_content(self):
 
     def _eval_subs(self, old, new):
         if old in self.variables:
```
Comment (asmeurer) [inline sympy/core/function.py]: Something that's not really handled correctly here is that `self.variables` can be any kind of expression, like `Subs(f(x), x*y, 0)`. I'm not completely sure what the correct logic should be, especially for multiple Subs (although I guess the best way to think about that is to think about them as nested Subs). 

But you don't need to fix this in this pull request if you don't want to. The fix here is already an improvement.

**Comment by asmeurer** [comment]
See my most recent comment. If you want to try to fix this so that it is 100% correct, go for it. Otherwise, I am +1 with what is here.

**Comment by certik** [comment]
I would merge this PR and then we can improve upon it.

@smichr I think the correct/most natural way to think is in the sense of this PR. It's also what other CASes seem to do, e.g. Sage:
```
sage: f(x).diff(x).subs({x: 0}).subs({x: y})
D[0](f)(0)
sage: f(x).diff(x).subs({x: 0})
D[0](f)(0)
```

**Pull request merged**